### PR TITLE
Fix #3169: ItemSelectionInteraction text change on selection

### LINKED ISSUE #3169: ItemSelectionInteraction text change on selection
**EDIT** (by @seanlip): See updated description here -- https://github.com/oppia/oppia-android/issues/3169#issuecomment-1225305069

----------------------

**Previous description**

[Mock1](https://xd.adobe.com/view/dfe8ccd4-39eb-414d-a0b2-b5674357f966-8abb/screen/722a0d0b-737d-4f07-95aa-0cfa7db06530/)

As per Mock1, by default the text is `Please select atleast one choice.`. If atleast one option is selected we hide this text and if the option is deselected then we show this text again.

Add this functionality and its relevant test.

## Expected Cases:

1. No item selected - Default Text - `Please select one or more choices.`
2. If total selection count is atleast 1 and less than allowable selections, change text to `You may select more choices.`
3. If total selection count is equal to allowable selections, change text to `No more than 4 choices may be selected.`

You can experiment similar cases on [What is a Fraction?](https://www.oppia.org/explore/umPkwp0L1M0-) . You will need to answer a lot of states before you will reach a state where you will find this interaction where you can select multiple options.

### PR DESCRIPTION
**Pull request by darkmat13r**
## Explanation
Fix #3169 : ItemSelectionInteraction text change on selection

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [ ] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by darkmat13r** [comment]
@rt4914 PTAL

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:1444]: All the test cases looks really nice.

Can you add one more test cases which basically tests following.
1. Select atleast 1 option.
2. Check that `Text` is gone.
3. Now unselect all the selected options.
4. Check that `Text` is visible again.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@darkmat13r Thanks a lot. Just need to add one more test case otherwise LGTM

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:1444]: @rt4914 New Test case added

Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:1444]: @rt4914 PTAL

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

@BenHenning Any idea why bazel tests are failing?

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: Can you add tests to demonstrate that the selection text retains its showing/not showing behavior after performing a rotation & activity recreation?

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: All but the selection bit can be removed since it's redundant with the earlier test to verify that the text hides when selected.

File: app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:58]: Please add a KDoc explaining what this value contains & when it changes. We don't have great KDoc coverage in view models, but we ought to try to moving forward (& especially now that we have a KDoc CI checker that's expecting full documentation coverage for all public API-facing methods & fields).

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (BenHenning) [inline app/src/main/res/layout/selection_interaction_item.xml:49]: Given the OR here, do we already have tests verifying that the text is always showing for radio buttons for both unselected & selected states to contrast with the checkbox behavior?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @darkmat13r! Just had a few comments--PTAL.

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (darkmat13r) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: @BenHenning  No, I will add them.

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: @BenHenning Yes. I will add the test case.

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (darkmat13r) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: @BenHenning For the radio buttons the text will always be hidden. I am not sure we need tests for this.

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: @BenHenning I have remove the landscape test cases as these doesn't make sense because currently when you select some answers and rotate the screen all values gets lost. So right now we can't add those test cases related to orientation change or activity recreation.
Follow are the removed test cases for landscape orientation:
`testStateFragment_interactions_land_noItemSelected_selectionTextIsDisplayed`
`testStateFragment_interactions_land_singleItemSelected_selectionTextIsNotDisplayed`
`testStateFragment_interactions_land_multipleItemSelected_selectionTextIsNotDisplayed`
`testStateFragment_interactions_changeConfiguration_selectUnselectCheckbox_selectionTextIsDisplayedCorrectly`

File: app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
Comment (darkmat13r) [reply app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:58]: @BenHenning I have added the KDoc for the variable.

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (darkmat13r) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: @BenHenning I have added test case for the radio buttons.

Comment (darkmat13r) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: @BenHenning PTAL

File: app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
Comment (darkmat13r) [reply app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:58]: @BenHenning PTAL

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: @BenHenning PTAL

**Comment by rt4914** [comment]
@darkmat13r Make sure you write something like
`PTAL @<username>` that way oppia-bot will assign the PR to correct reviewer.

**Comment by darkmat13r** [comment]
PTAL @BenHenning

**Comment by BenHenning** [comment]
Sorry, will need to review this tomorrow.

Also, closing & reopening to try and retrigger CI.

File: .idea/runConfigurations.xml
Comment (BenHenning) [inline .idea/runConfigurations.xml:1]: This is a change that's unrelated to your PR; please revert it.

File: gradle.properties
Comment (BenHenning) [inline gradle.properties:1]: These changes are unrelated to your PR; please revert them.

(If you need them locally, just make sure not to commit this file unless  you mean to change it).

File: app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt]:
```suggestion
  /** The number of currently selected answers. */
```

Rewording to make it a bit more concise.

File: app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:58]: Thanks--left one follow-up comment regarding the new comment.

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: Is this comment correct in this context? It seems we're verifying that the text _isn't_ shown in this case.

Ditto for the next test.

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:1392]: For this & the last test below, can we also verify that the text itself is correct?

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (BenHenning) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: I think you also need a test for state 2 (which is MultipleChoiceInput) to ensure that the text view is correctly arranged for that interaction, as well.

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: Ah, yes that's related to #1594. Fair point--I think omitting these tests are fine. Instead, can you add a TODO, such as:

```kotlin
// TODO(#1594): Add tests to ensure the selection text is properly showing/not showing after a configuration change.
```

**Review by BenHenning — COMMENTED** [review]
Thanks @darkmat13r! Just had a few folllow-up comments--PTAL.

File: app/src/main/res/layout/selection_interaction_item.xml
Comment (darkmat13r) [reply app/src/main/res/layout/selection_interaction_item.xml:49]: @BenHenning state 2 is Fraction input option. State 3 and state 4 are  multiple choice input. I have already added the test case for state 4.

```
@Test
 fun testStateFragment_interactions_radioItemSelection_selectionTextIsNotDisplayed() {
     launchForExploration(TEST_EXPLORATION_ID_2, shouldSavePartialProgress = false).use {
      startPlayingExploration()
      playThroughPrototypeState1()
      playThroughPrototypeState2()
      playThroughPrototypeState3()

      // Verify that the user is now on the fourth state.
      verifyViewTypeIsPresent(SELECTION_INTERACTION)
      onView(withId(R.id.selection_interaction_textview)).check(matches(not(isDisplayed())))
    }
  }

```
//Added these additional test cases
```
testStateFragment_interactions_singleRadioItemSelected_selectionTextIsNotDisplayed
testStateFragment_interactions_multipleRadioItemSelected_selectionTextIsNotDisplayed
testStateFragment_interactions_correctRadioItemSelected_selectionTextIsNotDisplayed
```

File: app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
Comment (darkmat13r) [reply app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt]: @BenHenning Yes, That is incorrect. Removed those comment. Same for below comment.

**Comment by rt4914** [comment]
@darkmat13r Can you please reply to all the open-comments, that way based on reply Ben can decide if that comment can be resolved or not. Thanks

**Comment by darkmat13r** [comment]
> @darkmat13r Can you please reply to all the open-comments, that way based on reply Ben can decide if that comment can be resolved or not. Thanks

@BenHenning  @rt4914 All the comments are resolved but I am not able to run the test cases. Some test cases are failing so I need to fix them.

**Comment by BenHenning** [comment]
FYI I'm listed as a codeowner for this PR & I'll be unavailable to perform code reviews over the next 2 weeks--thanks for your patience.

**Comment by BenHenning** [comment]
> > @darkmat13r Can you please reply to all the open-comments, that way based on reply Ben can decide if that comment can be resolved or not. Thanks
> 
> @BenHenning @rt4914 All the comments are resolved but I am not able to run the test cases. Some test cases are failing so I need to fix them.

@darkmat13r why aren't you able to run the test cases? It'd be really helpful to get more context. Please see https://github.com/oppia/oppia-android/wiki/Frequent-Errors-and-Solutions#facing-error-while-debugging-code which includes tips on debugging, and instructions for creating a debug doc that you can share with folks to make it easier for them to help you.

**Comment by darkmat13r** [comment]
@BenHenning  Not able to run the test cases on android studio artic fox version. The solution i mention to work on Java Idea community version also not working . I will provide more details on the error.
https://github.com/oppia/oppia-android/pull/3783

**Pull request closed without merging**
